### DISCUSSION TITLE
Checking to see if an application would be viable with Tauri

### DISCUSSION THREAD

**Discussion by andrewtavis** [tauri-apps/tauri#10403] [Q&A] [answered]
Hey all 👋

Sending along a message to check on the viability of building an application with Tauri. The [Scribe](https://github.com/scribe-org) community makes applications that focus on using open language data - predominately from [Wikidata](https://wikidata.org/) - to help language learners. [Scribe-iOS](https://github.com/scribe-org/Scribe-iOS) is our main end user application, which is a keyboard app that allows language learners to conjugate verbs, translate words and get reminded of noun genders from within the keyboard so that they don't need to migrate to another app or the web for help (among many features). We've had the idea for some time for [Scribe-Desktop](https://github.com/scribe-org/Scribe-Desktop), which we're recently thinking will be written in Rust instead of Python. 

The general idea for the application would be:
- There'd be a Scribe-Data GUI
- The GUI would be able to log the keystrokes of the user in any other application
- As the user types they can see annotations of the words like noun genders in the GUI
- Pressing a keyboard shortcut would switch the focus to the GUI and allow the user to type in verbs to conjugate, words to translate, etc with the commands then being executed

The general question to you all is: is this something that you all believe could be built with Tauri? Are there any limitations as far as GUIs not being able to read in activity from other applications or anything else we might not be thinking of that would make the feature set above impossible?

Thanks so much for any input you all can provide! Suggestions would also be more than welcome 😊

**Comment by FabianLars** [comment, marked as answer]
> The GUI would be able to log the keystrokes of the user in any other application
> Are there any limitations as far as GUIs not being able to read in activity from other applications

First of all, Tauri doesn't has anything built-in for that, but there are crates like [rdev](https://crates.io/crates/rdev) that do a pretty good job.

In general this is a bit platform dependent, while Windows shouldn't be a problem, Wayland users on Linux could be a problem, and macOS also has permission requirements (see rdev readme) but should work if granted.

The rest sounds like really ever GUI framework should be able to handle, Tauri included.

**Reply by andrewtavis** [reply]
Thanks so much for the quick answer here, @FabianLars :) Appreciate it a lot!

CC @ryanpaulgannon, we discussed recently whether giving Tauri another look so we can leverage Vue would be a good idea. Let me know if this sounds like it could be a bit easier, as it looks like your PR is also using rdev for the keyboard actions 😊

**Reply by RyanPaulGannon** [reply]
Nice one for this, @andrewtavis, and thanks for the info @FabianLars! 

Let me check this out next week, as you say, rdev is what we're using for the global listener. We just had the issue with the GUI and the listener both wanting to run on the main thread, but the listener won't work correctly due to security restraints on Mac on a secondary thread. I'll take another dive in to Tauri 🥸

**Reply by FabianLars** [reply]
rdev should work in a secondary thread too as long as it's the same process (which a normal thread is). If you're facing problems, please reach out.

**Comment by Shotman** [comment]
> `The GUI would be able to log the keystrokes of the user in any other application`

Sounds like a keylogger to me ... Personally I wouldn't install an app being able to record all my keystrokes ... Credit card info, password, etc etc
Even more if it's closed source. I don't know how antiviruses would react to such an app.

**Reply by andrewtavis** [reply]
100% would not be closed source, @Shotman, and thanks for bringing up the concern. And this is a good point that we'll need to keep in mind in terms of anti virus software.

**Reply by Shotman** [reply]
One thing that would be really nice would be to make the app 100% offline.
If the app can be run sandboxed completely cut off the internet, that would be even better and more reassuring to end users.

**Reply by andrewtavis** [reply]
This is a great point, @Shotman :) As of now Scribe applications are 100% offline, but we're thinking of doing data pack downloads so that a user doesn't need to have data from other languages on their devices. This is maybe more of a consideration for mobile apps and not for desktop where the size considerations wouldn't be as dire.

Tons of ways to go about this 🤔 Maybe data packs and updates for desktop could be bundled in similar ways such that by default the user needs to manually check for an update, or they can also switch it so the application will let them know when there's new data that's available? The general plan is that we're going to be running our data service every two weeks to get new language data from Wikidata, so maybe there's a "check for updates feature" that by default is more chronological than one with internet access that can literally check if there's something there if they enable it to do so. The default could be that data updates happen on say Sundays every two weeks, and then a few days later the Desktop app is coded to tell the user to check for an update? Might be a good way of going about this to assure that we're putting the best foot forward for privacy concerns.

Thanks again! Really appreciate it 😊

CC @wkyoshida

**Reply by andrewtavis** [reply]
Thinking about this further, another way we could do this would be that a person could build the Desktop application from source and then we could provide an option where they could download and install the data packs directly from the server project that we're working on. The command to install the project after cloning it from GitHub could include a --no-internet flag that would disable the applications ability to access the server or check for regular updates of the software itself.

We could have a "Remind me to update" option that would send a reminder every two weeks, or the user could also set the timespan after which they're reminded :)
